### PR TITLE
[Issue #26] Implement control for the repository format in format.yaml

### PR DESCRIPTION
**Pull request by lukasburda**
Issue: https://github.com/xstefank/tyr/issues/26

### CODE REVIEW

File: src/main/java/org/xstefank/webhook/WebHookEndpoint.java [hunk 1]
```diff
@@ -53,9 +55,18 @@ private static FormatConfig readConfig() {
...
 
         try {
-            return mapper.readValue(configFile, FormatConfig.class);
+            FormatConfig formatConfig = mapper.readValue(configFile, FormatConfig.class);
+            String repoFormat = formatConfig.getRepository();
+            if (repoFormat.matches("^[A-z,0-9].*/[A-z,0-9,-].*$") && repoFormat.split("/").length == 2) {
+                return formatConfig;
+            }
+            else
+                throw new BadRepositoryFormatException("Bad repository format in configuration file");
```
Comment (xstefank) [inline src/main/java/org/xstefank/webhook/WebHookEndpoint.java]: do you think this check should be here? what if more checks will be added in the future. It would be useful to define some kind of external validation

File: src/main/java/org/xstefank/control/Control.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.xstefank.control;
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public interface Control {
```
Comment (xstefank) [inline src/main/java/org/xstefank/control/Control.java]: can you rename this to Verification please

File: src/main/java/org/xstefank/control/ControlHandler.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package org.xstefank.control;
+
+import org.xstefank.model.yaml.FormatConfig;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ControlHandler {
```
Comment (xstefank) [inline src/main/java/org/xstefank/control/ControlHandler.java]: all control rename to verification, control is rather restrictive

File: src/main/java/org/xstefank/control/ControlHandler.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            String message = null;
+            for (Control control : controls) {
+                message = control.control(formatConfig);
+            }
+            if (message == null) {
+                return true;
+            } else {
+                throw new BadFormatException(message);
+            }
+        } catch (BadFormatException e) {
```
Comment (xstefank) [inline src/main/java/org/xstefank/control/ControlHandler.java]: you are catching exception you are throwing in try block

File: src/main/java/org/xstefank/control/RepositoryFormatControl.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+public class RepositoryFormatControl implements Control {
+
+    private final String REGEX = "^[A-z,0-9].*/[A-z,0-9,-].*$";
+    private final String MESSAGE = "Bad repository format in configuration file";
+
+    public String control(FormatConfig formatConfig) {
+        String repoFormat = formatConfig.getRepository();
+        if (repoFormat.matches(REGEX) && repoFormat.split("/").length == 2) {
+            return null;
```
Comment (xstefank) [inline src/main/java/org/xstefank/control/RepositoryFormatControl.java]: returning null is generally a bad practice, especially if it is undocumented, the caller of this method can easily get NPE

File: src/main/java/org/xstefank/webhook/WebHookEndpoint.java [hunk 6]
```diff
@@ -53,9 +57,12 @@ private static FormatConfig readConfig() {
         ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
 
         try {
-            return mapper.readValue(configFile, FormatConfig.class);
+            FormatConfig formatConfig = mapper.readValue(configFile, FormatConfig.class);
+            if (controlHandler.control(formatConfig))
```
Comment (xstefank) [inline src/main/java/org/xstefank/webhook/WebHookEndpoint.java]: `if (isValidConfiguration(formatConfig))`

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public class RepositoryFormatVerification implements Verification {
+
+
+    private final String REGEX = "^[A-z,0-9].*/[A-z,0-9,-].*$";
+    private final String MESSAGE = "Bad repository format in configuration file";
+
+    public boolean verify(FormatConfig formatConfig) {
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: please add `@Override` - https://stackoverflow.com/questions/94361/when-do-you-use-javas-override-annotation-and-why

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        String repoFormat = formatConfig.getRepository();
+        if (repoFormat.matches(REGEX) && repoFormat.split("/").length == 2)
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: this can still be abbreviated to one line

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+            return true;
+        return false;
+    }
+
+    public String getErrorMessage() {
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: `@Override`

File: src/main/java/org/xstefank/verification/Verification.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package org.xstefank.verification;
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public interface Verification {
+
+    boolean verify(FormatConfig formatConfig);
+    String getErrorMessage();
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/Verification.java]: it would be sufficient to document here how `getErrorMessage` behaves in case `verify` wasn't called or the verification was successful, maybe it would be better to ask if it does make sense as a separate method

File: src/main/java/org/xstefank/verification/VerificationHandler.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                continue;
+            else
+                throw new BadFormatException(verification.getErrorMessage());
+        }
+    }
+
+    private List<Verification> registerVerifications() {
+        List<Verification> verifications = new ArrayList<>();
+
+        verifications.add(new RepositoryFormatVerification());
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/VerificationHandler.java]: can you come up with one or two more verifications so all this effort is justified please?

File: src/main/java/org/xstefank/webhook/WebHookEndpoint.java [hunk 12]
```diff
@@ -7,21 +7,26 @@
...
 import java.io.File;
 import java.io.IOException;
+
 import static org.xstefank.check.TemplateChecker.TEMPLATE_FORMAT_FILE;
 
 @Path("/")
 public class WebHookEndpoint {
 
     private static final Logger log = Logger.getLogger(WebHookEndpoint.class);
+    private static VerificationHandler verificationHandler = new VerificationHandler();
```
Comment (xstefank) [inline src/main/java/org/xstefank/webhook/WebHookEndpoint.java]: I don't think it's required to keep this handler in a field as its only used in one method, which is (I may be wrong here) called only once on the following line

File: src/main/java/org/xstefank/verification/VerificationHandler.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    private List<Verification> verifications;
+
+    public VerificationHandler() {
+        verifications = registerVerifications();
+    }
+
+    public void verifyConfiguration(FormatConfig formatConfig) throws BadFormatException {
+        for (Verification verification : verifications) {
+            if (verification.verify(formatConfig))
+                continue;
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/VerificationHandler.java]: why is this necessary here?

File: src/main/java/org/xstefank/verification/EmptyFormatVerification.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public class EmptyFormatVerification implements Verification {
+
+    private final String MESSAGE = "Format in configuration file cannot be empty";
+
+    @Override
+    public String verify(FormatConfig formatConfig) {
+        return formatConfig.getFormat() != null ? "valid" : MESSAGE;
```
[the remark this answers is not in the document]
Comment (lukasburda) [reply src/main/java/org/xstefank/verification/EmptyFormatVerification.java]: You mean that i should directly verify the config file and not FormatConfig which loads the Mapper ? Am I right?

[the remark this answers is not in the document]
Comment (xstefank) [reply src/main/java/org/xstefank/verification/EmptyFormatVerification.java]: yes, however I think if you are able to declare the configuration without format element than this is a valid case, it would be worth while to check if the element can be made required somehow

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+public class RepositoryFormatVerification implements Verification {
+
+    private final String MESSAGE = "Bad repository format in configuration file";
+    private final String REGEX = "^[A-z,0-9]*./[A-z,0-9,-].*$";
+
+    @Override
+    public void verify(FormatConfig formatConfig) throws BadFormatException {
+        String repoFormat = formatConfig.getRepository();
+        if (!repoFormat.matches(REGEX) || repoFormat.split("/").length != 2) {
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: I am not sure `|| repoFormat.split("/").length != 2` is required

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+package org.xstefank.verification;
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public class RepositoryFormatVerification implements Verification {
+
+    private final String MESSAGE = "Bad repository format in configuration file";
+    private final String REGEX = "^[A-z,0-9]*./[A-z,0-9,-].*$";
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: are you sure this regex is right? can you please include the link to the official rules for the repository naming conventions/rules?

File: src/main/java/org/xstefank/webhook/WebHookEndpoint.java [hunk 17]
```diff
@@ -7,22 +7,26 @@
...
 import java.io.IOException;
+
 import static org.xstefank.check.TemplateChecker.TEMPLATE_FORMAT_FILE;
 
 @Path("/")
 public class WebHookEndpoint {
 
     private static final Logger log = Logger.getLogger(WebHookEndpoint.class);
-    private FormatConfig config = readConfig();
+    private static final FormatConfig config = readConfig();
```
Comment (xstefank) [inline src/main/java/org/xstefank/webhook/WebHookEndpoint.java]: why this change?

File: src/main/java/org/xstefank/webhook/WebHookEndpoint.java [hunk 18]
```diff
@@ -51,11 +55,12 @@ private static FormatConfig readConfig() {
         log.info(configFileName);
         File configFile = new File(configFileName);
         ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
-
         try {
-            return mapper.readValue(configFile, FormatConfig.class);
-        } catch (IOException e) {
+            FormatConfig formatConfig = mapper.readValue(configFile, FormatConfig.class);
+            new VerificationHandler(formatConfig).verifyConfiguration();
```
Comment (xstefank) [inline src/main/java/org/xstefank/webhook/WebHookEndpoint.java]: looking at this line alone, wouldn't it be better to have a static method on the Verification handler that takes formatConfig as an argument?

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java (hunk 15 above)
Comment (xstefank) [reply src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: @lukasburda try to type `test/slash-repo` as a name of the new repository at GitHub ;)

Comment (xstefank) [reply src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: but I see now that you want to check whether what the user typed is a valid repository, still I am not convinced that this check is necessary and it cannot be caught directly by regex

File: src/main/java/org/xstefank/verification/RepositoryFormatVerification.java [hunk 19]
```diff
@@ -0,0 +1,12 @@
+package org.xstefank.verification;
+
+import org.xstefank.model.yaml.FormatConfig;
+
+public class RepositoryFormatVerification implements Verification {
+
+    @Override
+    public void verify(FormatConfig formatConfig) throws IllegalConfigurationException {
+        if (!formatConfig.getRepository().matches("^[a-zA-Z0-9_]*./[a-zA-Z0-9_-]*.$"))
```
Comment (xstefank) [inline src/main/java/org/xstefank/verification/RepositoryFormatVerification.java]: what is the meaning of the dot character in front of the dollar sign please?

File: src/test/java/org/xstefank/verification/VerificationsTest.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
+package org.xstefank.verification;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.xstefank.model.yaml.FormatConfig;
+
+import static org.xstefank.TestUtils.*;
```
Comment (xstefank) [inline src/test/java/org/xstefank/verification/VerificationsTest.java]: please don't use wildcard imports, there should be check for this in checkstyle
